src/txn/txn_timestamp.c

```diff
@@ -668,6 +680,83 @@ __wt_txn_set_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
 	return (0);
 }
 
+/*
+ * __wt_txn_parse_prepare_timestamp --
+ *	Parse a request to set a transaction's prepare_timestamp.
+ */
+int
+__wt_txn_parse_prepare_timestamp(
+    WT_SESSION_IMPL *session, const char *cfg[], wt_timestamp_t *timestamp)
+{
+	WT_CONFIG_ITEM cval;
+
+	WT_RET(__wt_config_gets_def(session,
+	    cfg, "prepare_timestamp", 0, &cval));
+	if (cval.len > 0) {
+#ifdef HAVE_TIMESTAMPS
+		WT_TXN *prev;
+		WT_TXN_GLOBAL *txn_global;
+		wt_timestamp_t oldest_ts;
+		char hex_timestamp[2 * WT_TIMESTAMP_SIZE + 1];
+
+		txn_global = &S2C(session)->txn_global;
+
+		if (F_ISSET(&session->txn, WT_TXN_HAS_TS_COMMIT))
+			WT_RET_MSG(session, EINVAL,
+			    "commit timestamp should not have been set before "
+			    "prepare transaction");
+
+		WT_RET(__wt_txn_parse_timestamp(
+		    session, "prepare", timestamp, &cval));
+
+		/*
+		 * Prepare timestamp must be later/greater than all active read
+		 * timestamps.
+		 */
+		__wt_readlock(session, &txn_global->read_timestamp_rwlock);
+		prev = TAILQ_LAST(&txn_global->read_timestamph,
+		    __wt_txn_rts_qh);
+		if (prev != NULL &&
+		    __wt_timestamp_cmp(&prev->read_timestamp, timestamp) >= 0) {
+			__wt_readunlock(session,
+			    &txn_global->read_timestamp_rwlock);
+			WT_RET(__wt_timestamp_to_hex_string(session,
+			    hex_timestamp, &prev->read_timestamp));
+			WT_RET_MSG(session, EINVAL,
+			    "prepare timestamp %.*s not later than an active "
+			    "read timestamp %s ", (int)cval.len, cval.str,
+			    hex_timestamp);
+		}
+		__wt_readunlock(session, &txn_global->read_timestamp_rwlock);
+
+		/*
+		 * If there are no active readers, prepare timestamp must not
+		 * be older than oldest timestamp.
+		 */
+		if (prev == NULL) {
```

Comment:
If there are active readers, would we allow a prepare ts older that oldest ts?

Reply:
no active reader will be older than oldest ts. so if we validate against active reader, it implies we have already validated against oldest ts.